app/api/crud.py

```diff
@@ -195,13 +195,53 @@ async def get(
                             "subject_group": "first",
                             "assessment": lambda x: list(x.unique()),
                             "image_modal": lambda x: list(x.unique()),
-                            "pipeline_version": lambda x: list(x.unique()),
-                            "pipeline_name": lambda x: list(x.unique()),
                             "session_file_path": "first",
                         }
                     )
                 )
 
+                pipeline_data = (
+                    group.groupby(
+                        [
+                            "sub_id",
+                            "session_id",
+                            "session_type",
+                            "pipeline_name",
+                        ]
+                    )
+                    .agg(
+                        {
+                            "pipeline_version": lambda x: list(
+                                x.dropna().unique()
+                            )
+                        }
+                    )
+                    .reset_index()
+                )
+
+                pipeline_dict = (
+                    pipeline_data.groupby(
+                        ["sub_id", "session_id", "session_type"]
+                    )
+                    .apply(
+                        lambda x: dict(
+                            zip(x["pipeline_name"], x["pipeline_version"])
+                        )
+                    )
+                    .reset_index(name="completed_pipelines")
+                )
+
+                subject_data = pd.merge(
+                    subject_data.reset_index(drop=True),
+                    pipeline_dict,
+                    on=["sub_id", "session_id", "session_type"],
+                    how="left",
+                )
+
+                subject_data["completed_pipelines"] = subject_data[
```

Comment:
🍒 , for clarity on why this is needed:
```suggestion
                # ensure that for sessions missing completed pipeline info, completed_pipelines is still a dict rather than null/nan
                subject_data["completed_pipelines"] = subject_data[
```